### PR TITLE
Snap to grid

### LINKED ISSUE #11: Implement snap to grid

### PR DESCRIPTION
**Pull request by tlein**
Resolves #11

### CODE REVIEW

File: src/duckling/canvas/tools/_grid-snap.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    rightMiddle : Vector,
+    bottomMiddle : Vector,
+    leftMiddle : Vector 
+};
+
+/**
+ * Get the minimum snap distance for the given corners
+ */
+export function minCornerSnapDistance(movementDestinationWithoutSnap : Vector, box : Box2, gridSize : Vector) : Vector {
+    let cornersSnapDistance = _cornersSnapDistance(movementDestinationWithoutSnap, box, gridSize);
```
Comment (jeffswenson) [inline src/duckling/canvas/tools/_grid-snap.ts]: This code could be a lot simpler.

function minCornerSnapDistance(...) {
let snapDistance;
let minHypotenus = infinity;
for (let xDir of [-1, 0, 1]) {
    for (let yDir of [-1, 0, 1]) {
    let distance = _cornerSnapDistance({x: xDir, y: yDir}, desitnationWithoutSnap, box, gridSize);
    let hypotenus = _cornerSnapHypotenus(distance);
    if (hypotenus < minHypotenus) {
        minHypotenus = hypotenus;
        snapDistance = distance;
    }
}
return snapDistance;
}

File: src/duckling/canvas/tools/entity-move-tool.ts [hunk 2]
```diff
@@ -110,27 +117,30 @@ export class EntityMoveTool extends BaseTool {
...
         if (box) {
             graphics = new Graphics();
             graphics.lineStyle(1 / canvasZoom, 0x3355cc, 1);
             drawRectangle(box.position, box.dimension, graphics);
         }
         return graphics;
     }
+    
+    private _getSelectedEntityBox() {
+        let selectedEntityKey = this._selectionService.selection.value.selectedEntity;
```
Comment (jeffswenson) [inline src/duckling/canvas/tools/entity-move-tool.ts]: These two lines could be simplified  to this._selectionService.selection.value.entity;

File: src/duckling/canvas/tools/entity-move-tool.ts [hunk 3]
```diff
@@ -52,13 +54,18 @@ export class EntityMoveTool extends BaseTool {
...
     onStageMove(event : CanvasMouseEvent) {
         if (this._selection) {
-            this._entityPositionService.setPosition(
-                this._selection,
-                vectorRound(vectorAdd(event.stageCoords, this._selectOffsetCoords)),
-                this._mergeKey);
+            let destination = vectorRound(vectorAdd(event.stageCoords, this._selectOffsetCoords));
+            let gridSize = this._projectService.project.getValue().currentMap.gridSize;
+            if (!this._isSnapToGrid(event)) {
+                let gridSize = this._projectService.project.getValue().currentMap.gridSize;
```
Comment (jeffswenson) [inline src/duckling/canvas/tools/entity-move-tool.ts]: This is a duplicate.
let gridSize = this._projectService.project.getValue().currentMap.gridSize;

File: src/duckling/canvas/tools/entity-move-tool.ts [hunk 4]
```diff
@@ -52,13 +54,18 @@ export class EntityMoveTool extends BaseTool {
...
+                destination = vectorSubtract(destination, minCornerSnapDistance(
```
Comment (jeffswenson) [inline src/duckling/canvas/tools/entity-move-tool.ts]: This should be vector add. That will allow you to remove the random negation in minCornerSnapDistance.

**Review by jeffswenson — CHANGES_REQUESTED** [review]

File: src/duckling/canvas/tools/_grid-snap.ts [hunk 5]
```diff
@@ -1,88 +1,28 @@
...
-        return cornersSnapDistance.middle;
-    } else if (minHypotenus === topMiddleHypotenus) {
-        return cornersSnapDistance.topMiddle;
-    } else if (minHypotenus === rightMiddleHypotenus) {
-        return cornersSnapDistance.rightMiddle;
-    } else if (minHypotenus === bottomMiddleHypotenus) {
-        return cornersSnapDistance.bottomMiddle;
-    } else if (minHypotenus === leftMiddleHypotenus) {
-        return cornersSnapDistance.leftMiddle;
+    let minSnap = {x : gridSize.x, y : gridSize.y};
```
Comment (tlein) [inline src/duckling/canvas/tools/_grid-snap.ts]: minSnap is not used

File: src/duckling/canvas/tools/entity-move-tool.ts (hunk 4 above)
Comment (tlein) [reply src/duckling/canvas/tools/entity-move-tool.ts]: fixed

File: src/duckling/canvas/tools/_grid-snap.ts (hunk 5 above)
Comment (tlein) [reply src/duckling/canvas/tools/_grid-snap.ts]: fixed

File: src/duckling/canvas/tools/entity-move-tool.ts (hunk 3 above)
Comment (tlein) [reply src/duckling/canvas/tools/entity-move-tool.ts]: fixed

File: src/duckling/canvas/tools/entity-move-tool.ts (hunk 2 above)
Comment (tlein) [reply src/duckling/canvas/tools/entity-move-tool.ts]: fixed

**Review by jeffswenson — APPROVED** [review]

**Pull request merged**